### DISCUSSION TITLE
tsconfig.json path alias not working

### DISCUSSION THREAD

**Discussion by silviubogan** [jaredpalmer/razzle#1637] [Q&A] [answered]
The tsconfig.json is this:

```ts
{
  "compilerOptions": {
    "target": "ES6",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "noFallthroughCasesInSwitch": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    "jsx": "react-jsx",
    "strictNullChecks": true,
    "skipLibCheck": true,
    "baseUrl": "src",
    "paths": {
      "@/*": ["./*"]
    }
  },
  "include": ["src"]
}
```

The part

    "paths": {
      "@/*": ["./*"]
    }

does not function in my case.

This is in my package.json file:
`"start": "BROWSER=chromium concurrently \"npm run start:tsc\" \"razzle start\"",`

I run npm start and I get this message in the section of client compilation errors:

`Cannot find module '@/redux/authSlice'`.

The dependencies I think are relevant are these:

    "razzle": "4.0.4",
    "razzle-dev-utils": "4.0.4",

I also tried putting the following in my `razzle.config.js` file but it is without success:

```js
const defaultModify = ({
  env: { target, dev },
  webpackConfig: config,
  webpackObject: webpack,
}) => {
  config.resolve.alias = {
    "@/": "src/",
  };
  return config;
};

module.exports = {
  modifyWebpackConfig: ({
    env: { target, dev },
    webpackConfig,
    webpackObject,
  }) => {
    const defaultConfig = defaultModify({
      env: { target, dev },
      webpackConfig,
      webpackObject,
    });
    return defaultConfig;
  },
};
```

What would be the next step in debugging this issue? How can I solve this issue?

Thanks!

**Comment by fivethreeo** [comment]
What is your apps directory structure. Also see https://github.com/jaredpalmer/razzle/blob/master/packages/razzle/config/modules.js#L53

**Reply by fivethreeo** [reply]
Did you solve it?

**Reply by silviubogan** [reply]
Not yet. Thanks for the answer.

**Reply by silviubogan** [reply]
The app directory structure:
![image](https://user-images.githubusercontent.com/198924/122651797-e3cfea00-d143-11eb-81c0-d3d5c17b709d.png)

**Reply by fivethreeo** [reply]
Try

    "paths": {
      "@/*": ["../*"]
    }

**Reply by fivethreeo** [reply]
Um, maybe not 

    "paths": {
      "@/*": ["../src/*"]
    }

**Reply by silviubogan** [reply]
Thanks! I've tried both and the old error is gone but, for the first of your code snippets I get hundreds of errors of this form:

> Cannot find module '@/types' or its corresponding type declarations.

Also, the server is not compiling too because I have the same error as in my first post in this topic.

**Reply by fivethreeo** [reply]
Hm, not sure why this does not work. Can you try just running tsc to see if this is a razzle or tsc issue?

**Reply by silviubogan** [reply]
> Hm, not sure why this does not work. Can you try just running tsc to see if this is a razzle or tsc issue?

I tried running `npx tsc -b` (as in the build script, `tsc -b && razzle build`, and I get lots of errors like this:

> Module '"redux"' has no exported member 'combineReducers'

which make no sense to me (combineReducers is in the module redux).

If I use one of the code snippets you provided, I get errors that modules starting with `@/` are not found. Without your changes, I get errors like the above one.

Maybe the problem is also related to the baseUrl option? I have this: `"baseUrl": "src",`.

Maybe it is relevant that I am migrating this project from react-app-rewired to Razzle.

Thanks!

**Reply by silviubogan** [reply]
> What is your apps directory structure. Also see https://github.com/jaredpalmer/razzle/blob/master/packages/razzle/config/modules.js#L53

The two occurences of the following string in the code link received from you:

```js
replace(/\/\*$/, '')
```

I think they break my tsconfig.json if the strings on which they operate are retrieved from tsconfig.json.

How can I debug my project together with the source code of Razzle?

**Reply by fivethreeo** [reply]
https://razzlejs.org/docs/customization#options

Try setting the config debug flag

**Reply by fivethreeo** [reply]
Razzle does not modify tsconfig.json it only reads values from it. But it may be that tsc treats paths differently somehow or that resolve.alias for webpack is generated wrong by razzle somehow, see what debug prints and show it here.

**Reply by silviubogan** [reply]
@fivethreeo 

I thought it would be a good idea to do this first. The first command that is executed by npm start is this:

```shell
$ tsc -b -w --preserveWatchOutput
```

The output of just this command is, without the [...] masks and the SomeFile placeholder, the following:

```log
12:47:09 - Starting compilation in watch mode...

[...] error TS2339: Property 'getState' does not exist on type 'EnhancedStore<any, any, Middlewares<any>>'.
[...] error TS2305: Module '"redux"' has no exported member 'combineReducers'.
[...] error TS2339: Property 'getState' does not exist on type 'EnhancedStore<any, any, any[]>'.
[...] error TS7006: Parameter 'x' implicitly has an 'any' type.
[...] error TS2724: 'React' has no exported member named 'ForwardedRef'. Did you mean 'forwardRef'?
[...] error TS2724: 'React' has no exported member named 'ForwardedRef'. Did you mean 'forwardRef'?
[...] error TS2724: 'React' has no exported member named 'ForwardedRef'. Did you mean 'forwardRef'?
[...] error TS2724: 'React' has no exported member named 'ForwardedRef'. Did you mean 'forwardRef'?
[...] error TS2724: 'React' has no exported member named 'ForwardedRef'. Did you mean 'forwardRef'?
[...] error TS2724: 'React' has no exported member named 'ForwardedRef'. Did you mean 'forwardRef'?
[...] error TS2724: 'React' has no exported member named 'ForwardedRef'. Did you mean 'forwardRef'?
[...] error TS7006: Parameter 'x' implicitly has an 'any' type.
[...] error TS7006: Parameter 'x' implicitly has an 'any' type.
[...] error TS2724: 'React' has no exported member named 'ForwardedRef'. Did you mean 'forwardRef'?
[...] error TS2724: '"react"' has no exported member named 'ForwardedRef'. Did you mean 'forwardRef'?
[...] error TS7006: Parameter 'x' implicitly has an 'any' type.
[...] error TS7006: Parameter 'x' implicitly has an 'any' type.
[...] error TS2305: Module '"redux"' has no exported member 'Dispatch'.
[...] error TS7006: Parameter 'x' implicitly has an 'any' type.
[...] error TS7006: Parameter 'x' implicitly has an 'any' type.
[...] error TS7006: Parameter 'x' implicitly has an 'any' type.
[...] error TS1208: 'SomeFile.test.tsx' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.
[...] error TS2305: Module '"redux"' has no exported member 'Reducer'.
[...] error TS2305: Module '"redux"' has no exported member 'Action'.
[...] error TS2339: Property 'type' does not exist on type 'ActionWithPayload<any, any>'.

[...]

12:47:19 - Found 30 errors. Watching for file changes.
```

I will try the debug flag and post here what I think it is relevant (sorry for the masks above, but it is not an open-source project, I guess respected the good practice).

Thanks.

**Reply by silviubogan** [reply]
With debug.options set to true I get the following output but on line 25057 (the one with lots of `****`) I have the same output as in the previous message.

https://gist.github.com/silviubogan/2995e76138984a73abd83f6c11c600a7

Next, I think I will try to set the other variables in the debug section, one by one.

**Reply by silviubogan** [reply]
With debug.config enabled I get this:

https://gist.github.com/silviubogan/edc01119489f59908e90413e07d35902

In the same gist, here is the same output offered by tsc in [my preceding code snippet](https://github.com/jaredpalmer/razzle/discussions/1637#discussioncomment-896876):

https://gist.github.com/silviubogan/edc01119489f59908e90413e07d35902#file-npm-start-2-log-L13994

**Reply by silviubogan** [reply]
With debug.nodeExternals option set I get the same output in the client compilation errors section but in the server compilation errors I get lots of lines like:

> [1] Not externalizing /path/to/my/project/node_modules/razzle-dev-utils/prettyNodeErrors.js (relative require)
> [1] Not externalizing /path/to/my/project/node_modules/webpack/hot/poll.js?300 (relative require)
> [1] Not externalizing /path/to/my/project/src (relative require)
> [1] Not externalizing !!/path/to/my/project/node_modules/razzle-start-server-webpack-plugin/dist/monitor-loader.js!/path/to/my/project/node_modules/razzle-start-server-webpack-plugin/dist/monitor-loader.js (cannot resolve)
> [1] Not externalizing fs (default)
> [1] Not externalizing jest-message-util (real resolution differs)
> [1] Externalizing @babel/code-frame (node_modules)
> [1] Not externalizing ./log (relative require)
> [1] Not externalizing ./log-apply-result (relative require)
> [1] Externalizing webpack/hot/log-apply-result (node_modules)
> [1] Not externalizing ./log (relative require)
> [1] Not externalizing path (default)
> [1] Externalizing graceful-fs (node_modules)
> [1] Externalizing chalk (node_modules)
> [1] Not externalizing micromatch (real resolution differs)
> [1] Externalizing slash (node_modules)
> [1] Externalizing @babel/code-frame (node_modules)
> [1] Externalizing stack-utils (node_modules)
> [1] Externalizing express (node_modules)
> [1] Not externalizing ./server (relative require)
> [1] Not externalizing util (default)
> [1] Not externalizing braces (real resolution differs)
> [1] Externalizing picomatch (node_modules)

And lots of them have file names that are inside the code files I work on so I wouldn't give their names if there isn't a strong reason for this.

At the end of the output there is this:

> [...]
> [1] Not externalizing @/config/[...] (cannot resolve)
> [1] ✔ Server: Compiled with some errors in 2.31s
> [1] redux-persist failed to create sync storage. falling back to noop storage.
> [1] /path/to/my/project/node_modules/react-syntax-highlighter/dist/esm/styles/prism/index.js:1
> [1] export { default as coy } from './coy';
> [1] ^^^^^^
> [1] 
> [1] SyntaxError: Unexpected token 'export'
> [1]     at compileFunction (<anonymous>)
> [1]     at wrapSafe (internal/modules/cjs/loader.js:1054:16)
> [1]     at Module._compile (internal/modules/cjs/loader.js:1102:27)
> [1]     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
> [1]     at Module.load (internal/modules/cjs/loader.js:986:32)
> [1]     at Function.Module._load (internal/modules/cjs/loader.js:879:14)
> [1]     at Module.require (internal/modules/cjs/loader.js:1026:19)
> [1]     at require (internal/modules/cjs/helpers.js:72:18)
> [1]     at Object.react-syntax-highlighter/dist/esm/styles/prism (/path/to/my/project/external "react-syntax-highlighter/dist/esm/styles/prism":1:1)
> [1]     at __webpack_require__ (/path/to/my/project/build/webpack:/webpack/bootstrap:754:1)
> [1] sswp> !!! Script exited with code 1
> [0] npm run start:tsc exited with code SIGINT
> [1] razzle start exited with code SIGINT

**Comment by fivethreeo** [comment]
```js
// razzle.config.js
'use strict';

module.exports = {
  modifyWebpackOptions({
    env: {
      target, // the target 'node' or 'web'
      dev, // is this a development build? true or false
    }
    options: {
      webpackOptions, // the default options that will be used to configure webpack/ webpack loaders and plugins
    }
  }) {
    webpackOptions.notNodeExternalResMatch = (request, context) => {
       return /react-syntax-highlighter/.test(request)
    };
    webpackOptions.babelRule.include = webpackOptions.babelRule.include.concat([
      /react-syntax-highlighter/
    ]);
    return webpackOptions;
  }
};
```

**Reply by fivethreeo** [reply]
The ts errors comes from tsc, so may be some options you need to add to tsconfig and maybe fix some types and the mentioned forwardRef react import

"noImplicitAny": false

**Reply by fivethreeo** [reply]
https://stackoverflow.com/questions/56119468/module-file-path-react-redux-has-no-exported-member-dispatch - may be relevant

**Reply by silviubogan** [reply]
> The ts errors comes from tsc, so may be some options you need to add to tsconfig and maybe fix some types and the mentioned forwardRef react import
> 
> "noImplicitAny": false

It is interesting because I had the same tsconfig.json file when the app was just a react-app-rewired project, just separated into two: tsconfig.json which extended tsconfig.base.json. I still have those original files in the main branch of the app (I created another branch for migrating to Razzle).

I know it may seem ugly, but I am considering to use Next.js if Razzle does not work.

I am looking into the last 3 messages you posted.

**Comment by fivethreeo** [comment]
Theese are all fixable issues imo :)

On Mon, 21 Jun 2021 at 20:09, Silviu Bogan ***@***.***> wrote:

> The ts errors comes from tsc, so may be some options you need to add to
> tsconfig and maybe fix some types and the mentioned forwardRef react import
>
> "noImplicitAny": false
>
> It is interesting because I had the same tsconfig.json file when the app
> was just a react-app-rewired project, just separated into two:
> tsconfig.json which extended tsconfig.base.json. I still have those
> original files in the main branch of the app (I created another branch for
> migrating to Razzle).
>
> I know it may seem ugly, but I am considering to use Next.js if Razzle
> does not work.
>
> I am looking into the last 3 messages you posted.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jaredpalmer/razzle/discussions/1637#discussioncomment-899378>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAADGXLXURBHA3ZRFFQEQTTT5543ANCNFSM467AF6SQ>
> .
>


-- 
Øyvind Saltvik

**Reply by silviubogan** [reply]
Thanks for all the help!

The current razzle.config.js is this:

```js
// razzle.config.js
"use strict";

module.exports = {
  plugins: [
    {
      name: "scss",
      options: {
        // postcss: {
        //   dev: {
        //     sourceMap: true,
        //   },
        // },
        css: {
          dev: {
            sourceMap: true,
            importLoaders: 1,
            modules: {
              auto: true,
              localIdentName: "[name]__[local]___[hash:base64:5]",
            },
          },
          prod: {
            sourceMap: true /* razzleOptions.enableSourceMaps */,
            importLoaders: 1,
            modules: {
              auto: true,
              localIdentName: "[name]__[local]___[hash:base64:5]",
            },
          },
        },
      },
    },
  ],
  module: {
    rules: [
      {
        test: /\.css$/i,
        use: ["style-loader", "css-loader"],
      },
      {
        test: /\.scss$/i,
        use: ["style-loader", "css-loader"],
      },
    ],
  },
  modifyWebpackOptions({
    env: {
      target, // the target 'node' or 'web'
      dev, // is this a development build? true or false
    },
    options: {
      webpackOptions, // the default options that will be used to configure webpack/ webpack loaders and plugins
    },
  }) {
    webpackOptions.notNodeExternalResMatch = (request, context) => {
      return /react-syntax-highlighter/.test(request);
    };
    webpackOptions.babelRule.include = webpackOptions.babelRule.include.concat([
      /react-syntax-highlighter/,
    ]);
    return webpackOptions;
  },
};
```

The issue (besides the tsc issues) is that the CSS files are not being read properly:

![image](https://user-images.githubusercontent.com/198924/122810441-0a268e80-d2d8-11eb-9044-2f4deeb66aa2.png)

**Reply by silviubogan** [reply]
Without having read the entire Razzle docs it is hard for me to understand this page: https://razzlejs.org/plugins/razzle-plugin-scss#css.

**Comment by fivethreeo** [comment]
https://razzlejs.org/getting-started#common-issues

On Mon, 21 Jun 2021 at 20:34, Silviu Bogan ***@***.***> wrote:

> Without having read the entire Razzle docs it is hard for me to understand
> this page: https://razzlejs.org/plugins/razzle-plugin-scss#css.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jaredpalmer/razzle/discussions/1637#discussioncomment-899474>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAADGSBKVQQHUEEYRAHZS3TT6A5BANCNFSM467AF6SQ>
> .
>


-- 
Øyvind Saltvik

**Reply by silviubogan** [reply]
Thanks! I think I'll look into it tomorrow. Also, congratulations for working on Razzle!

**Reply by fivethreeo** [reply]
Thanks :)

**Reply by fivethreeo** [reply]
```js
// razzle.config.js
"use strict";

module.exports = {
  plugins: ["scss"],
  modifyWebpackOptions({
    env: {
      target, // the target 'node' or 'web'
      dev, // is this a development build? true or false
    },
    options: {
      webpackOptions, // the default options that will be used to configure webpack/ webpack loaders and plugins
    },
  }) {
    webpackOptions.notNodeExternalResMatch = (request, context) => {
      return /react-syntax-highlighter|react-virtualized/.test(request);
    };
    webpackOptions.babelRule.include = webpackOptions.babelRule.include.concat([
      /react-syntax-highlighter/,
      /react-virtualized/
    ]);
    return webpackOptions;
  },
};
```

**Reply by silviubogan** [reply]
I tried with:

```json
"noImplicitAny": false,
```

and with the code in https://github.com/jaredpalmer/razzle/discussions/1637#discussioncomment-899572.

I am still trying to figure out what are the mistakes in https://github.com/jaredpalmer/razzle/discussions/1637#discussioncomment-899463 and to understand why I get so many errors from `tsc`.

**Reply by fivethreeo** [reply]
But some of the errors went away?

**Reply by silviubogan** [reply]
> But some of the errors went away?

Yes, from about 30 tsc errors I have just 22 now. Thanks!

**Reply by fivethreeo** [reply]
What about strict, try false there aswell

**Reply by fivethreeo** [reply]
or it might be react and redux versions

**Reply by silviubogan** [reply]
> What about strict, try false there aswell

With it false I get 34 errors instead of just 22.

> or it might be react and redux versions

Yes. Most of them are about different versions of some packages, from what I've seen.

I just posted [this](https://github.com/reduxjs/redux-toolkit/discussions/1204), it is the first error that I am getting and other errors are based on it. I think I should go through them from top to bottom.

**Reply by silviubogan** [reply]
@fivethreeo 

I took a look at https://github.com/jaredpalmer/razzle/blob/master/packages/razzle-plugin-scss/index.js.

I don't know what I am missing from `razzle.config.js`:

```js
module.exports = {
  plugins: ["scss"],
  module: {
    rules: [
      {
        test: /\.css$/i,
        use: ["css-loader"],
      },
      {
        test: /\.scss$/i,
        use: ["scss-loader"],
      },
    ],
  },
  modifyWebpackOptions({ env: { target, dev }, options: { webpackOptions } }) {
    webpackOptions.notNodeExternalResMatch = (request, context) => {
      return /react-syntax-highlighter|react-virtualized/.test(request);
    };
    webpackOptions.babelRule.include = webpackOptions.babelRule.include.concat([
      /react-syntax-highlighter/,
      /react-virtualized/,
    ]);
    return webpackOptions;
  },
};
```

I still get the error:

```log
/path/to/my/project/node_modules/react-virtualized/styles.css:3
.ReactVirtualized__Collection {
^

SyntaxError: Unexpected token '.'
    at compileFunction (<anonymous>)
    at wrapSafe (internal/modules/cjs/loader.js:1054:16)
    at Module._compile (internal/modules/cjs/loader.js:1102:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
    at Module.load (internal/modules/cjs/loader.js:986:32)
    at Function.Module._load (internal/modules/cjs/loader.js:879:14)
    at Module.require (internal/modules/cjs/loader.js:1026:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.<anonymous> (/path/to/my/project/node_modules/react-select-virtualized/dist/index.min.js:1:331)
    at Module._compile (internal/modules/cjs/loader.js:1138:30)
sswp> !!! Script exited with code 1
```

Thanks.

**Reply by fivethreeo** [reply]
react-virtualized is still externalized, this razzle.config.js should fix that

**Reply by fivethreeo** [reply]
module in razzle.config.js does nothing by the way

**Reply by fivethreeo** [reply]
ah, i see now react-select-virtualized

**Reply by fivethreeo** [reply]
add that to regexes aswell :)

**Reply by silviubogan** [reply]
If I try to run npm start now, I get this error:

```
[1] /path/to/my/project/node_modules/tiny-slider/src/tiny-slider.js:23
[1] import { raf } from './helpers/raf.js';
[1] ^^^^^^
[1] 
[1] SyntaxError: Cannot use import statement outside a module
[1]     at compileFunction (<anonymous>)
[1]     at wrapSafe (internal/modules/cjs/loader.js:1001:16)
[1]     at Module._compile (internal/modules/cjs/loader.js:1049:27)
[1]     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
[1]     at Module.load (internal/modules/cjs/loader.js:950:32)
[1]     at Function.Module._load (internal/modules/cjs/loader.js:790:14)
[1]     at Module.require (internal/modules/cjs/loader.js:974:19)
[1]     at require (internal/modules/cjs/helpers.js:92:18)
[1]     at Object.<anonymous> (/path/to/my/project/node_modules/tiny-slider-react/build/Carousel.js:12:19)
[1]     at Module._compile (internal/modules/cjs/loader.js:1085:14)
[1] sswp> !!! Script exited with code 1
```

If I set in package.json type: module, then I get this:

```
 ERROR  Invalid razzle.config.js file.
[1] 
[1] Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /path/to/my/project/razzle.config.js
[1] require() of ES modules is not supported.
[1] require() of /path/to/my/project/razzle.config.js from /path/to/my/project/node_modules/razzle/config/loadRazzleConfig.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
[1] Instead rename razzle.config.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /path/to/my/project/package.json.
[1] 
[1]     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1102:13)
[1]     at Module.load (internal/modules/cjs/loader.js:950:32)
[1]     at Function.Module._load (internal/modules/cjs/loader.js:790:14)
[1]     at Module.require (internal/modules/cjs/loader.js:974:19)
[1]     at require (internal/modules/cjs/helpers.js:92:18)
[1]     at /path/to/my/project/node_modules/razzle/config/loadRazzleConfig.js:23:18
[1]     at new Promise (<anonymous>)
[1]     at module.exports (/path/to/my/project/node_modules/razzle/config/loadRazzleConfig.js:12:10)
[1]     at /path/to/my/project/node_modules/razzle/scripts/start.js:43:5
[1]     at new Promise (<anonymous>)
[1]     at main (/path/to/my/project/node_modules/razzle/scripts/start.js:42:10)
[1]     at Object.<anonymous> (/path/to/my/project/node_modules/razzle/scripts/start.js:223:1)
[1]     at Module._compile (internal/modules/cjs/loader.js:1085:14)
[1]     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
[1]     at Module.load (internal/modules/cjs/loader.js:950:32)
[1]     at Function.Module._load (internal/modules/cjs/loader.js:790:14) {
[1]   code: 'ERR_REQUIRE_ESM'
[1] }
[1] razzle start exited with code 1
```

I am not sure what is the next step in debugging this issue.

**Reply by silviubogan** [reply]
I renamed razzle.config.js to razzle.config.cjs and now I get:

```log
[1] ℹ Compiling Server
[1] ✔ Server: Compiled with some errors in 1.93s
[1] ReferenceError: module is not defined in ES module scope
[1] This file is being treated as an ES module because it has a '.js' file extension and '/path/to/my/project/package.json' contains "type": "module". To treat it as a CommonJS script, rename it to use the '.cjs' file extension.
[1]     at file:///path/to/my/project/build/server.js:1:1
[1]     at ModuleJob.run (internal/modules/esm/module_job.js:169:25)
[1]     at Loader.import (internal/modules/esm/loader.js:177:24)
[1]     at Object.loadESM (internal/process/esm_loader.js:68:5)
[1] sswp> !!! Script exited with code 1
```

**Reply by silviubogan** [reply]
I guess I should just rename the build/server.js file to build/server.cjs but I don't know how to do it and it looks ugly to me. I never did this before.

**Reply by fivethreeo** [reply]
https://nodejs.org/api/packages.html#packages_determining_module_system

**Reply by fivethreeo** [reply]
And add tiny-slider to regexes aswell

**Reply by silviubogan** [reply]
1. tiny-slider.js is not treated as an ESM and it should be.
2. I make it an ESM by putting in my project's package.json "type": "module"
3. razzle.config.js, being an ESM, must be loaded with `import`, but it is with `require`
4. I rename razzle.config.js to razzle.config.cjs so that it is treated not as an ESM but as CJS
5. the problem is that the built bundle file, `/path/to/my/project/build/server.js` has the first line as `module.exports = ...` which is CJS

I put these in tsconfig.json:

    "target": "ES2021",
    "module": "ES2020",

and it is still not working.

I do not know how to change the output file's target if not in tsconfig.json as I tried above by setting `target` or `module` options.

**Comment by fivethreeo** [comment]
1. it must be transpiled to cjs hence the regex
2. don't ;)
3. not if 2 is not done
4. same
5. same

**Reply by silviubogan** [reply]
Thanks. Now I get this error:

```log
[1] /path/to/my/project/node_modules/razzle-plugin-scss/node_modules/sass-loader/dist/index.js:30
[1]   (0, _schemaUtils.validate)(_options.default, options, {
[1]                             ^
[1] 
[1] Module build failed (from ./node_modules/razzle-plugin-scss/node_modules/sass-loader/dist/cjs.js):
[1] TypeError: (0 , _schemaUtils.validate) is not a function
[1] 
[1]   59217 | /***/ (function(module, exports) {
[1]   59218 |
[1] > 59219 | throw new Error("Module build failed (from ./node_modules/razzle-plugin-scss/node_modules/sass-loader/dist/cjs.js):\nTypeError: (0 , _schemaUtils.validate) is not a function\n    at Object.loader (/path/to/my/project/node_modules/razzle-plugin-scss/node_modules/sass-loader/dist/index.js:30:29)");
[1]   59220 |
[1]   59221 | /***/ }),
[1]   59222 |
[1] 
[1]     at Object.loader (/path/to/my/project/node_modules/razzle-plugin-scss/node_modules/sass-loader/dist/index.js:30:29)
[1]     at Object../src/components/SomeComponent/SomeComponent.scss (/path/to/my/project/build/server.js:59219:7)
[1]     at __webpack_require__ (/path/to/my/project/webpack/bootstrap:754:1)
[1]     at fn (/path/to/my/project/build/webpack:/webpack/bootstrap:59:1)
[1]     at Module../src/components/SomeComponent/SomeComponent.tsx (/path/to/my/project/build/server.js:59236:81)
[1]     at __webpack_require__ (/path/to/my/project/build/webpack:/webpack/bootstrap:754:1)
[1]     at fn (/path/to/my/project/build/webpack:/webpack/bootstrap:59:1)
[1]     at Module../src/components/SomeComponent/index.tsx (/path/to/my/project/build/webpack:/src/components/SomeComponent/index.tsx:1:1)
[1]     at __webpack_require__ (/path/to/my/project/build/webpack:/webpack/bootstrap:754:1)
[1]     at fn (/path/to/my/project/build/webpack:/webpack/bootstrap:59:1)
[1]     at Module../src/components/index.tsx (/path/to/my/project/build/webpack:/src/components/index.tsx:1:1)
[1] 
[1] sswp> !!! Script exited with code 1
```

and in the SomeComponent/index.tsx file there is just a single line of code:

```tsx
export { SomeComponent } from "./SomeComponent";
```

**Comment by fivethreeo** [comment]
I think you have conflicting module names. Scss tries to import the tsx.

**Reply by silviubogan** [reply]
I have a directory SomeComponent that contains SomeComponent.scss, SomeComponent.tsx and index.tsx. SomeComponent.scss is used directly by SomeComponent.tsx and index.tsx just exports SomeComponent from SomeComponent.tsx.

But I renamed SomeComponent.scss to style.scss and I get the same error.

I also tried putting `declare module '*.scss';` inside a .d.ts file with the same result.

**Reply by fivethreeo** [reply]
What does style.scss contain?

**Reply by silviubogan** [reply]
```scss
.dzu-dropzone {
  overflow: hidden;
}

.dzu-previewFileName {
  word-break: break-all;
  width: auto;
}
```

**Reply by fivethreeo** [reply]
Did you remove SomeComponent.scss, and the error is the same?

**Comment by fivethreeo** [comment]
Um, It is the other way around ;) Scss tries to load tsx.

**Reply by silviubogan** [reply]
If I remove the scss file and the statement that imports it, the error moves its eyes onto the next scss file in my project directories, giving me the same error message just with the file name changed.

**Reply by fivethreeo** [reply]
I read that traceback wrong it is a error coming from the loader trying to load the scss. I think you may have some wrong dependencies somewhere

**Reply by fivethreeo** [reply]
schema-utils might be it

**Reply by silviubogan** [reply]
Thanks. I updated typescript and razzle-plugin-scss to their latest versions and moved them into devDependencies in package.json and now I get this error inside tiny-slider package:

```
[1] /path/to/my/project/node_modules/tiny-slider/src/helpers/raf.js:1
[1] var win = window;
[1] ^
[1] ReferenceError: window is not defined
```

I think I should declare window somewhere, maybe in razzle.config.js. I am going to search on web about this.

**Reply by silviubogan** [reply]
I found this:

Using Client-Side Only Packages | Gatsby
https://www.gatsbyjs.com/docs/using-client-side-only-packages/

through [this](https://github.com/jechav/tiny-slider-react/issues/25).

**Comment by fivethreeo** [comment, marked as answer]
Answer to the original question. Was conflicting module names.
